pints/_mcmc/__init__.py

```diff
@@ -300,22 +309,65 @@ class MCMCController(object):
     """
 
     def __init__(
-            self, log_pdf, chains, x0, sigma0=None, transform=None,
+            self, log_pdf, chains, x0=None, sigma0=None, transform=None,
             method=None):
 
         # Check function
         if not isinstance(log_pdf, pints.LogPDF):
             raise ValueError('Given function must extend pints.LogPDF')
 
+        # Get number of parameters
+        self._n_parameters = log_pdf.n_parameters()
+
+        # Check number of chains
+        self._n_chains = int(chains)
+        if self._n_chains < 1:
+            raise ValueError('Number of chains must be at least 1.')
+
+        # Get functions which can generate initial position or set initial
+        # position points if numerical values given
+        self._x0_isfunction = False
+        if x0 is None:
+            # use prior if one's available
+            if isinstance(log_pdf, pints.LogPosterior):
+                x0 = log_pdf.log_prior()
+            # do what Stan does and initialise all parameters on U(-2, 2) scale
+            else:
+                x0 = pints.UniformLogPrior([-2] * self._n_parameters,
+                                           [2] * self._n_parameters)
+        if isinstance(x0, pints.LogPrior):
+            init = [x0.sample() for i in range(chains)]
+            self._init_fn = x0
+            self._x0_isfunction = True
+        else:
+            if len(x0) != chains:
+                raise ValueError(
+                    'Number of initial positions must be equal to number of'
+                    ' chains.')
+            if not all([len(x) == self._n_parameters for x in x0]):
+                raise ValueError(
+                    'All initial positions must have the same dimension as the'
+                    ' given LogPDF.')
+            init = x0
+
         # Apply a transformation (if given). From this point onward the MCMC
         # sampler will see only the transformed search space and will know
         # nothing about the model parameter space.
         if transform is not None:
             # Convert log pdf
             log_pdf = transform.convert_log_pdf(log_pdf)
 
+        # Store transform for later detransformation: if using a transform, any
+        # parameters logged to the filesystem or printed to screen should be
+        # detransformed first!
+        self._transform = transform
+
+        # Store function
+        self._log_pdf = log_pdf
+
+        if transform is not None:
             # Convert initial positions
-            x0 = [transform.to_search(x) for x in x0]
+            init = [transform.to_search(x) for x in init]
```

Comment:
Another transformation thing to check! I'd guess we want to sample from the transformed log prior, so that this conversion is unnecessary - so this `to_search` should happen only for an explicitly set x0 ?

Reply:
This is a good point but I just checked: doing log_pdf = transform.convert_log_pdf(log_pdf) transforms the pdf but does not transform the prior itself (indeed, as far as I can tell, a LogPosterior converted to a TransformedLogPDF does not have its own TransformedLogPrior). So, I think something like this needs to happen.

Reply:
@chonlei can you look at these last two transformation-related points? The easiest place is not here but on the "Files changed" tab.

Reply:
For the first point, the convert_log_pdf() only considers the PDF as one thing (it doesn't care if it's a PDF/likelihood/posterior). There is the TransformedLogPrior object which can be called via convert_log_prior() to do the conversion first(?), then we could sample from there (but note that there's no way that we can sample directly from the transformed prior easily, so it samples from the model/original parameter space and does the transform afterward).

Reply:
But I see Michael's point that we've now shifted things to "transforming parameters in the middle of the code" which is something that i was trying to avoid. I don't see the code will break things at the moment, but I somehow wanted to have transformation happens only at the start and end, so that the middle bit doesn't involve any transformation in case of (future) confusion...